include/mxnet/op_attr_types.h

```diff
@@ -30,12 +92,34 @@ using nnvm::NodeAttrs;
  *
  *  \note Register under "FCreateLayerOp"
  */
-using FCreateLayerOp = std::function<
-  Operator* (const NodeAttrs& n,
-             Context ctx,
-             const std::vector<TShape>& in_shape,
-             const std::vector<int>& in_type)>;
-
+using FCreateOpState = std::function<dmlc::any (const NodeAttrs& attrs,
+                                                Context ctx,
+                                                const std::vector<TShape>& in_shape,
+                                                const std::vector<int>& in_type)>;
+/*!
+ * \brief Execution mode of this operator.
+ */
+using FExecType = std::function<ExecType (const NodeAttrs& attrs)>;
+/*!
+ * \brief Resiger a compute function for stateful operator.
+ *
+ * \note Register under "FStatefulCompute<cpu>" and "FStatefulCompute<gpu>"
+ */
+using FStatefulCompute = std::function<void (const dmlc::any& state,
```

Comment:
state it's self cannot be mutated. If it's a pointer the content can be mutated.
The thing is state can be copied around and you cannot modify it inplace.

Another option is to use `shared_ptr<any> state` but that forces you to use a pointer and you cannot use make_shared to construct state inplace.